### DISCUSSION TITLE
why does WalkBuilder::filter_entry() require a `'static` closure?

### DISCUSSION THREAD

**Discussion by matthiasbeyer** [BurntSushi/ripgrep#2633] [Q&A] [answered]
https://github.com/BurntSushi/ripgrep/blob/7099e174acbcbd940f57e4ab4913fee4040c826e/crates/ignore/src/walk.rs#L899

This argument restriction seems weirdly restrictive to me. Is there any particular reason why the closure one has to pass in has to be a pure function (`Fn + 'static`)? Any chance that these might get lifted?

**Comment by BurntSushi** [comment, marked as answer]
It doesn't have to be a pure function. The `'static` just means it can't borrow anything. That's very different from what a "pure" function is.

I don't have the specific context, but you might consider changing the signature to what you want and seeing what breaks. Consider, for example, that [spawning a thread has a similar restriction](https://doc.rust-lang.org/std/thread/fn.spawn.html).

**Reply by matthiasbeyer** [reply]
>  The 'static just means it can't borrow anything. That's very different from what a "pure" function is.

You're absolutely right of course... bad wording on my side.

Thanks for your fast reply and sorry that I didn't open this as a discussion in the first place. :heart:
